lib/HTFeed/JobMetrics.pm

```diff
@@ -0,0 +1,330 @@
+	    $jm->clear();
+	    print "JobMetrics cleared!\n";
+	} elsif ($arg =~ /^--observe$/) {
+	    # pass --observe to call observe(metric, value)
+	    # using the next 2 args as metric and value
+	    my $metric = shift @ARGV;
+	    my $value  = shift @ARGV;
+	    $jm->observe($metric, $value);
+	} elsif ($arg =~ /^--add$/) {
+	    # pass --add to call add(metric, value)
+	    # using the next 2 args as metric and value
+	    my $metric = shift @ARGV;
+	    my $value  = shift @ARGV;
+	    $jm->add($metric, $value);
+	} elsif ($arg =~ /^--inc$/) {
+	    # pass --inc to call inc(metric)
+	    # using the next arg as metric
+	    my $metric = shift @ARGV;
+	    $jm->inc($metric);
+	} else {
+	    print "Ignoring invalid arg: $arg\n";
+	}
+	# Now check the next arg.
+    }
+}
+# end commandline stuff
+
+sub new {
+    my $class = shift;
+    my $self  = {};
+
+    # TODO: Need to determine best place to store metrics
+    # pod's /tmp? /htprep? some other location?
+    $self->{file} = "/tmp/HTFeed::JobMetrics::Data";
+    if (defined $ENV{'HTFEED_JOBMETRICS_DATA_DIR'}) {
+	$self->{file} = $ENV{'HTFEED_JOBMETRICS_DATA_DIR'};
+    }
+    $self->{prom} = Prometheus::Tiny::Shared->new(
+	filename => $self->{file}
+    );
+
+    bless($self, $class);
+    $self->_setup_metrics();
+
+    return $self;
+}
+
+# Show where data is stored.
+sub loc {
+    my $self = shift;
+    $self->{file};
+}
+
+# for counters, where each observation is an arbitrary +=,
+# e.g. $jm->add("Download_ms", 219.5);
+# or $ perl lib/HTFeed/JobMetrics.pm --add Download_ms 219.5
+sub add {
+    my $self   = shift;
+    my $metric = shift;
+    my $value  = shift;
+
+    $self->_valid_metric($metric) && $self->{prom}->add($metric, $value);
+}
+
+# for integer counters, where each observation is a single tick ++
+# e.g. $jm->inc("Unpack_items");
+# or $ perl JobMetrics.pm --inc Unpack_items
+sub inc {
+    my $self   = shift;
+    my $metric = shift;
+
+    $self->_valid_metric($metric) && $self->{prom}->inc($metric);
+}
+
+# for histograms
+# e.g. $jm->observe("Unpack_ms", 125);
+# or $ perl JobMetrics.pm --observe Unpack_ms 125
+# Note that histograms must be set up with buckets ahead of time.
+sub observe {
+    my $self   = shift;
+    my $metric = shift;
+    my $value  = shift;
+
+    $self->_valid_metric($metric) && $self->{prom}->histogram_observe($metric, $value);
+}
+
+# just show names of metrics
+sub list_metrics {
+    my $self = shift;
+    [sort keys %{$self->{metrics}}];
+}
+
+# pretty-printable tabular data
+sub pretty {
+    my $self = shift;
+    $self->{prom}->format;
+}
+
+# Return metrics matching $pat
+sub match {
+    my $self = shift;
+    my $pat  = shift;
+    my @pretty = split("\n", $self->pretty);
+    [grep { /$pat/ } @pretty];
+}
+
+# irrevocably delete job metrics
+sub clear {
+    my $self = shift;
+    $self->{prom}->clear;
+}
+
+# We'll mostly get sizes in b (from perl's -s operator),
+# and want to show them in kb.
+sub b_to_kb {
```

Comment:
Prometheus suggests recording metrics as bytes rather than kilobytes; may be worth reviewing:

https://prometheus.io/docs/practices/naming/

Reply:
Alright, that simplifies some things.

Reply:
How do you feel about seconds/milliseconds though?

... that page:
> should use base units (e.g. seconds, bytes, meters - not milliseconds, megabytes, kilometers)

Reply:
I would use seconds and report a decimal/floating point value